tls/s2n_handshake_io.c

```diff
@@ -852,6 +852,20 @@ message_type_t s2n_conn_get_current_message_type(struct s2n_connection *conn)
     return ACTIVE_MESSAGE(conn);
 }
 
+static S2N_RESULT s2n_execute_handler(struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(conn);
+
+    /* Ensure the state machine referenced is consistent throughout the handshake */
+    if (conn->actual_protocol_version_established || IS_HELLO_RETRY_HANDSHAKE(conn) || WITH_EARLY_DATA(conn)) {
```

Comment:
I know you said that you were struggling to write it, but it seems like we really need an early data test since we've gotten this wrong for early data a couple times now.